design/baremetal-operator/bmh_boot_iso.md

```diff
@@ -0,0 +1,128 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add boot-iso API to BareMetalHost
+
+## Status
+
+provisional
+
+## Summary
+
+Add a new interface so that it is possible to boot arbitrary
+iso images via Ironic when using redfish-virtualmedia.
+
+## Motivation
+
+In some circumstances it is desirable to boot an existing iso image
+instead of having Ironic boot IPA, for example:
+
+To reduce boot time for ephemeral workloads, it may be possible to boot an iso
+and not deploy any image to disk (saving the time to write the image and reboot)
+
+Where an alternative installer exists that is distributed as a live-iso, it
+may be desirable to leverage that toolchain instead of the IPA deploy ramdisk,
+for example the [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/)
+
+An [Ironic API](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk) exists for this, but it's not currently accessible via metal3, so it would be desirable to optionally enable this workflow.
+
+### Goals
+
+Expose the Ironic API to boot iso images via metal3.
+
+### Non-Goals
+
+This only considers the case where the redfish-virtualmedia BMC
+option is used, not any ability to boot arbitrary iso images via
+PXE.
+
+## Proposal
+
+Add a new field to the BareMetalHost spec which indicates that instead of
+deploying an image to disk, a live-iso image will be booted:
+
+```
+  spec:
+    image:
```

Comment:
The `image` struct has several other fields related to each other. Would we use those? For example, do we need to verify the checksum of the ISO?

Could we trigger this different behavior if the `status.image.diskFormat` field is set to `ISO`? Or do we already support writing `ISO` images to disk?

Reply:
Currently the boot_iso interface Ironic exposes does not require a checksum and I've not checked if the redfish attachment API supports that @dtantsur any idea?  Given that Ironic doesn't require it and metal3 won't download the image to verify any checksum, I omitted that for now.

I don't believe that Ironic has any support for writing ISO images to disk, so we probably could use `status.image.diskFormat` provided we expect that to remain the case?

Reply:
OK, what I'm thinking about here is the fact that the fields under `spec.image` are all currently closely related. This new field is sort of related, but it's also sort of standalone because it sounds like the user would either set `spec.image.iso` or all of the other fields (with `diskFormat` remaining optional).

So, if the ISO URL isn't going to use the other image settings, perhaps it should be in its own group. You mention "live boot" in another comment, so how about `spec.liveBoot.url`? That would be parallel to the `spec.image.url` field, and give us a place to put any other settings we might need related to managing the live booting process. For example, if we decide to support PXE booting then we could have `spec.liveBoot.kernelURL` and `spec.liveBoot.ramDiskURL`.

Reply:
@dhellmann thanks for the feedback - `spec.liveBoot.*` sounds good to me - do you think it'd be reasonable to rework this proposal so we cover both the live-iso and PXE ramdisk cases supported by ironic?

Reply:
I could go either way on the PXE support. If we actually need it or want to support it now, then I could see including it in this proposal. If we don't need it right now, we could mention it as future work to explain why the API change is designed as it is and then have a separate enhancement later when we're ready to actually support it.

Reply:
Ok I've mentioned that we can boot an iso via iPXE but that this proposal doesn't consider kernel/ramdisk boot, only an iso